spec/mongoid/attributes_spec.rb

```diff
@@ -2684,4 +2684,18 @@
       end
     end
   end
+
+  context "when modifiying a hash referenced with the [] notation" do
+    let(:church) { Church.create!(location: { x: 1 }) }
+
+    before do
+      church[:location].merge!(y: 2)
```

Comment:
The idea behind this change, is that the following test would work on master if this was `church.location`, i.e. using the dot notation instead of `[]`. This happens because, the dot notation uses the field getter, which contains the following code: 
```
attribute_will_change!(name) if value.resizable?
```
which marks the hash as changed right before it changes it. The `[]` notation however, does not contain that line. Therefore, the hash is not marked as changed before the update, and the change to the hash is not persisted. This PR adds the above line of code to the read_attribute method.